stdlib/public/core/StringUnicodeScalarView.swift

```diff
@@ -308,123 +278,178 @@ extension String.UnicodeScalarView : RangeReplaceableCollection {
+  /// Returns the length of the first extended grapheme cluster in UTF-16
+  /// code units.
+  @warn_unused_result
+  @inline(never)
+  internal func _measureExtendedGraphemeClusterForward(
+    from start: Index
+  ) -> Int {
+    var end = self.endIndex
+    if start == end {
+      return 0
+    }
 
-    // Detect positions that have no corresponding index.
-    if !utf8Index._isOnUnicodeScalarBoundary {
-      return nil
+    let clusterBreakProperty = _UnicodeGraphemeClusterBreakPropertyTrie()
+    let clusterSegmenter = _UnicodeExtendedGraphemeClusterSegmenter()
+    var index = start
+
+    var gcb0 = clusterBreakProperty.getPropertyRawValue(
+      self[index].value
+    )
+
+    _nextInPlace(&index)
+
+    while index != end {
+      // FIXME(performance): consider removing this "fast path".  A branch
+      // that is hard to predict could be worse for performance than a few
+      // loads from cache to fetch the property 'gcb1'.
+      if clusterSegmenter.isBoundaryAfter(gcb0) {
+        break
+      }
+      let gcb1 = clusterBreakProperty.getPropertyRawValue(
+        self[index].value
+      )
+      if clusterSegmenter.isBoundary(gcb0, gcb1) {
+        break
+      }
+      gcb0 = gcb1
+      _nextInPlace(&index)
     }
-    self.init(utf8Index._coreIndex, core)
-  }
 
-  /// Construct the position in `unicodeScalars` that corresponds
-  /// exactly to `characterIndex`.
-  ///
-  /// - Precondition: `characterIndex` is an element of
-  ///   `String(unicodeScalars).indices`.
-  public init(
-    _ characterIndex: String.Index,
-    within unicodeScalars: String.UnicodeScalarView
-  ) {
-    self.init(characterIndex._base._position, unicodeScalars._core)
+    return index._position - start._position
   }
 
-  /// Returns the position in `utf8` that corresponds exactly
-  /// to `self`.
-  ///
-  /// - Precondition: `self` is an element of `String(utf8)!.indices`.
+  /// Returns the length of the previous extended grapheme cluster in UTF-16
+  /// code units.
   @warn_unused_result
-  public func samePosition(in utf8: String.UTF8View) -> String.UTF8View.Index {
-    return String.UTF8View.Index(self, within: utf8)
-  }
+  @inline(never)
+  internal func _measureExtendedGraphemeClusterBackward(
+    from end: Index
+  ) -> Int {
+    let start = self.startIndex
+    if start == end {
+      return 0
+    }
 
-  /// Returns the position in `utf16` that corresponds exactly
-  /// to `self`.
-  ///
-  /// - Precondition: `self` is an element of `String(utf16)!.indices`.
-  @warn_unused_result
-  public func samePosition(
-    in utf16: String.UTF16View
-  ) -> String.UTF16View.Index {
-    return String.UTF16View.Index(self, within: utf16)
-  }
+    let clusterBreakProperty = _UnicodeGraphemeClusterBreakPropertyTrie()
+    let clusterSegmenter = _UnicodeExtendedGraphemeClusterSegmenter()
+    var index = end
+
+    _previousInPlace(&index)
+    var gcb0 = clusterBreakProperty.getPropertyRawValue(
+      self[index].value
+    )
```

Comment:
I don't think that we put closing parentheses on a separate line in function calls.  Could you fold them back to the previous line?  (Here and elsewhere in this PR.)